File: leetcode/c++/2841_maximum_sum_of_almost_unique_subarray.cpp
// https://leetcode.cn/problems/maximum-sum-of-almost-unique-subarray

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long maxSum(vector<int>& nums, int m, int k) {
        unordered_map<int, int> array;
        long long res = 0, sum = 0;
        for (int l = 0, r = 0; r < nums.size(); r++) {
            array[nums[r]]++;
            sum += nums[r];
            if (r >= l + k) {
                if (--array[nums[l]] == 0)
                    array.erase(nums[l]);
                sum -= nums[l++];
            }
            if (r == l + k -1 && array.size() >= m) 
                res = max(res, sum);
        }
        return res;
    }
};

// 时间复杂度O(n)，空间复杂度O(n)

File: leetcode/golang/2841_maximum_sum_of_almost_unique_subarray.go
func max (a, b int64) int64 {
    if a > b {
        return a
    }
    return b
}

func maxSum(nums []int, m int, k int) int64 {
    array := make(map[int]int)
    res, sum := int64(0), int64(0)
    for l, r := 0, 0; r < len(nums); r++ {
        array[nums[r]]++
        sum += int64(nums[r])
        if r >= l + k {
            array[nums[l]]--
            if array[nums[l]] == 0 {
                delete(array, nums[l])
            }
            sum -= int64(nums[l])
            l++
        }
        if r == l + k - 1 && len(array) >= m {
            res = max(res, sum)
        }
    }
    return res
}